pkg/kubectl/cmd/exec/exec.go

```diff
@@ -266,11 +292,40 @@ func (o *StreamOptions) SetupTTY() term.TTY {
 
 // Run executes a validated remote execution against a pod.
 func (p *ExecOptions) Run() error {
-	pod, err := p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
-	if err != nil {
-		return err
+	var err error
+	// we still need legacy pod getter when PodName in ExecOptions struct is provided,
+	// since there are any other command run this function by providing Podname with PodsGetter
+	// and without resource builder, eg: `kubectl cp`.
+	if len(p.PodName) != 0 {
+		p.Pod, err = p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		builder := p.Builder().
+			WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
+			NamespaceParam(p.Namespace).DefaultNamespace()
+
+		switch len(p.Resources) {
```

Comment:
The whole switch statement is redundant, see how [ResourceNames](https://github.com/kubernetes/cli-runtime/blob/cdb522bcb34cfd041c0f6473be10c804e45eb0b3/pkg/resource/builder.go#L398) is implemented. The first argument should be "pods", ie. the default resource, and the second argument should be `argsIn[0]` (without splitting it).

Reply:
Same with the answer to your comment above.a

I did this to addressing soltysh comment about passing the actual type and name.

Reply:
See my reply [above](https://github.com/kubernetes/kubernetes/pull/73664/files#r276767421). All in all, I think what you had initially is the right approach.

Reply:
I think so.